File: Primes.cpp
#include "Primes.h"

Primes::Primes(void)
	: number(1)
{
}

Primes::~Primes(void)
{
}

bool Primes::isPrime(PrimeDataType number)
{
	// 1 isn't prime by definiton.
	if (2 > number)
	{
		return false;
	}

	//  Even numbers aren't prime too.
	if (0 == number % 2)
	{
		return false;
	}

	PrimeDataType divisor = 3;
	PrimeDataType divisible = 0;
	PrimeDataType rootOfNumber = static_cast<PrimeDataType>(round(sqrt(number)));

	do
	{
		if (0 == number % divisor)
		{
			divisible = 1;
		}
		divisor++;
	} while ((0 == divisible) && (rootOfNumber > divisor));

	if (divisible > 0) {
		return false;
	}

	return true;
}

PrimeDataType Primes::nextPrime(PrimeDataType number)
{
	PrimeDataType prime = 0;
	PrimeDataType i = number + 1;

	do
	{
		if (this->isPrime(i))
		{
			prime = i;
		}
		i++;
	} while (prime == 0);

	return prime;
}

PrimeDataType Primes::previousPrime(PrimeDataType number)
{
	PrimeDataType prime = 0;
	PrimeDataType i = number - 1;

	do
	{
		if (this->isPrime(i))
		{
			prime = i;
		}
		i--;
	} while (prime == 0);

	return prime;
}

File: Primes.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Primes
{
    class Primes
    {
        private const int DIVISION_START_NUMBER = 3;

        public bool isPrime(ulong number)
        {
            ulong divisor;
            ulong sqrtOfNumber;
            Boolean divisible;

            // 1 is not prime by definition.
            if (number < 2) {
                return false;
            }

            if (2 == number) {
                return true;
            }

            // even numbers are also not prime.
            if (0 == number % 2) {
                return false;
            }

            divisor = DIVISION_START_NUMBER;
            divisible = false;
            sqrtOfNumber = (ulong)Math.Ceiling(Math.Sqrt(number));

            while (divisor <= sqrtOfNumber) {
                if (0 == number % divisor) {
                    divisible = true;
                    break;
                }

                divisor++;
            };

            return !divisible;
        }

        public ulong nextPrime(ulong number)
        {
            ulong nextPrime = number + 1;

            while (!isPrime(nextPrime)) {
                nextPrime++;
            }

            return nextPrime;
        }

        public ulong prevPrime(ulong number)
        {
            ulong prevPrime = number;
            Boolean prevPrimeFound = isPrime(prevPrime);

            while (!prevPrimeFound) {
                prevPrime--;
                prevPrimeFound = isPrime(prevPrime);
            };

            return prevPrime;
        }
    }
}
